crates/iceberg/src/expr/visitors/manifest_evaluator.rs

```diff
@@ -0,0 +1,459 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::expr::visitors::bound_predicate_visitor::{visit, BoundPredicateVisitor};
+use crate::expr::visitors::inclusive_projection::InclusiveProjection;
+use crate::expr::{Bind, BoundPredicate, BoundReference};
+use crate::spec::{Datum, FieldSummary, ManifestFile, PartitionSpecRef, Schema, SchemaRef};
+use crate::{Error, ErrorKind};
+use fnv::FnvHashSet;
+use std::sync::Arc;
+
+/// Evaluates [`ManifestFile`]s to see if their partition summary matches a provided
+/// [`BoundPredicate`]. Used by [`TableScan`] to filter down the list of [`ManifestFile`]s
+/// in which data might be found that matches the TableScan's filter.
+pub(crate) struct ManifestEvaluator {
+    partition_schema: SchemaRef,
+    partition_filter: BoundPredicate,
+    case_sensitive: bool,
+}
+
+impl ManifestEvaluator {
+    pub(crate) fn new(
+        partition_spec: PartitionSpecRef,
+        table_schema: SchemaRef,
+        filter: BoundPredicate,
+        case_sensitive: bool,
+    ) -> crate::Result<Self> {
+        let partition_type = partition_spec.partition_type(&table_schema)?;
+
+        // this is needed as SchemaBuilder.with_fields expects an iterator over
+        // Arc<NestedField> rather than &Arc<NestedField>
+        let cloned_partition_fields: Vec<_> =
+            partition_type.fields().iter().map(Arc::clone).collect();
+
+        let partition_schema = Schema::builder()
+            .with_schema_id(partition_spec.spec_id)
+            .with_fields(cloned_partition_fields)
+            .build()?;
+
+        let partition_schema_ref = Arc::new(partition_schema);
+
+        let mut inclusive_projection = InclusiveProjection::new(partition_spec.clone());
+        let unbound_partition_filter = inclusive_projection.project(&filter)?;
+
+        let partition_filter = unbound_partition_filter
+            .rewrite_not()
+            .bind(partition_schema_ref.clone(), case_sensitive)?;
+
+        Ok(Self {
+            partition_schema: partition_schema_ref,
+            partition_filter,
+            case_sensitive,
+        })
+    }
+
+    /// Evaluate this `ManifestEvaluator`'s filter predicate against the
+    /// provided [`ManifestFile`]'s partitions. Used by [`TableScan`] to
+    /// see if this `ManifestFile` could possibly contain data that matches
+    /// the scan's filter.
+    pub(crate) fn eval(&self, manifest_file: &ManifestFile) -> crate::Result<bool> {
+        if manifest_file.partitions.is_empty() {
+            return Ok(true);
+        }
+
+        if self.partition_schema.schema_id() != manifest_file.partition_spec_id {
```

Comment:
I would suggest to add a comment to explain why comparing schema_id against partition_spec_id, or store original spec id directly, but it's not a blocker.

Reply:
Done. 👍🏼

Reply:
As mentioned above, likely, this doesn't match. We need to create an evaluator for each `partition_spec_id` that we see.

Reply:
https://github.com/apache/iceberg-rust/pull/323/files#diff-bbfbe5e334be6c501ba2ca0ddd84d658ff0f3f84f2b5b532212f4e096a09e09bR205-R207